### PR TITLE
merge serpapi-ruby from dev into master

### PR DESCRIPTION
**Pull request by jvmvik**
This dev branch includes the fully rewritten ruby client with one example per search engine supported.

The code quality has improved.
- lint clean using `rake lint`
- test passing with 100% code coverage `rake test`
- better README.md (generated from README.md.erb because the code snippets are extracted from the tests).

The only un-validated part is the release process itself: `rake release`
 The release process requires to setup an account with ruby gems.org. 
 I will performed the final release when this branch has been merged into dev.

At that point of time, I'd like to keep the feature set really simple.
Please review the Ruby code, flow and see if you can find something weird and suggest an improvement.

Thanks you'll

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,475 @@
...
+```bash
+$ gem install serpapi
+```
+
+[Link to the gem page](https://rubygems.org/gems/serpapi/)
+
+Tested Ruby versions:
+ - 2.2
+ - 2.5.2
+ - 3.0.0
```
Comment (ilyazub) [inline README.md:24]: Ruby versions in `ci.yml`: 3.0.3, 2.5.2, 2.7.5.

File: README.md [hunk 2]
```diff
@@ -1 +1,475 @@
...
+it prints your account information.
+
+## Basic example per search engine
+### Search Google Images
+```ruby
+require 'serpapi'
+client = SerpApi::Client.new(tbm: 'isch', api_key: ENV['API_KEY'], engine: 'google')
+image_results_list = client.search(q: 'coffee')[:images_results]
+image_results_list[0..2].each do |image_result|
+puts image_result[:original]
```
Comment (ilyazub) [inline README.md:165]:
```suggestion
image_results_list[0..2].each do |image_result|
  puts image_result[:original]
end
```

File: README.md [hunk 3]
```diff
@@ -1 +1,475 @@
...
+To download the image on Linux you could do: `wget #{image_result[:original]}`
+
+### Search youtube
+```ruby
+require 'serpapi'
+client = SerpApi::Client.new(api_key: ENV['API_KEY'])
+query = {
+engine: 'youtube',
+search_query: 'Coffee'
+}
```
Comment (ilyazub) [inline README.md:179]:
```suggestion
query = {
  engine: 'youtube',
  search_query: 'Coffee'
}
```

File: README.md.erb [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+### Search google events
+<%= snippet('ruby', 'spec/serpapi/example_search_google_events_spec.rb', 2, 6) %>
+(see https://serpapi.com/google_events)[https://serpapi.com/google_events]
+
+### Search google local services
+<%= snippet('ruby', 'spec/serpapi/example_search_google_local_services_spec.rb', 2, 6) %>
+(see https://serpapi.com/google_local_services)[https://serpapi.com/google_local_services]
+
+### Search google maps
+<%= snippet('ruby', 'spec/serpapi/example_search_google_maps_spec.rb', 2, 6) %>
```
Comment (ilyazub) [inline README.md.erb:238]: Great job on templating code examples :+1:

File: spec/serpapi/search_archive_api_spec.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  it 'fetch a search result from the archive' do
+    # mock
+    search_response_mock = '{
+        "search_metadata": {
+          "id": "cdbdbdebeabab1beb"
+        }
+      }'
+    client = SerpApi::Client.new(api_key: ENV['API_KEY'], engine: 'google')
+    if client.api_key.nil?
+      allow(search).to receive(:get_results) { search_response_mock }
```
Comment (ilyazub) [inline spec/serpapi/search_archive_api_spec.rb:13]: Is `search` defined?

**Review by ilyazub — DISMISSED** [review]
@jvmvik Thanks for unifying the SerpApi wrapper for Ruby :+1: 

Tests couldn't be executed on CI.

[Error](https://github.com/serpapi/serpapi-ruby/actions/runs/2096203312)

```
actions/checkout@v2 and ruby/setup-ruby@v1 are not allowed to be used in serpapi/serpapi-ruby. Actions in this workflow must be: within a repository owned by serpapi.
```

image

File: spec/serpapi/search_archive_api_spec.rb (hunk 5 above)
Comment (jvmvik) [reply spec/serpapi/search_archive_api_spec.rb:13]: good catch. forgot to rename search to client

File: README.md [hunk 6]
```diff
@@ -1 +1,475 @@
...
+```
+
+[Link to the gem page](https://rubygems.org/gems/serpapi/)
+
+Tested Ruby versions:
+ - 2.2
+ - 2.5.2
+ - 3.0.0
+
+see: Git hub actions.
```
[the remark this answers is not in the document]
Comment (jvmvik) [reply README.md]: fixed.

[the remark this answers is not in the document]
Comment (jvmvik) [reply README.md]: fixed by f886b6ea31f52f028e3b9ddddb4bf7b205bb9da0

File: README.md [hunk 7]
```diff
@@ -1 +1,475 @@
...
+  :keys=>["austin", "tx", "texas", "united", "states"]},
+  ...]
+```
+
+### Search Archive API
+
+This API allows to retrieve previous client.
+To do so run a client to save a search_id.
+```ruby
+client = SerpApi::Client.new(api_key: ENV['API_KEY'], engine: 'google')
```
[the remark this answers is not in the document]
Comment (jvmvik) [reply README.md:128]: fixed by 4d279bbe2f0c760c3dc7291772f81a3388f3a837

File: README.md (hunk 2 above)
Comment (jvmvik) [reply README.md]: code update to `pp results[:images_results]`. so it's more generic for all the templated spec files

**Comment by jvmvik** [comment]
I had to comment 2 tests because the backend is unstable at the moment:
- spec/serpapi/example_search_google_jobs_spec.rb
- spec/serpapi/example_search_google_play_spec.rb

Thanks @ilyazub for the detailed review.
@hartator if you give us the green light, I can release the gem.

**Comment by jvmvik** [comment]
actually there are one more issue with yard documentation where the Markdown source does not render properly.

**Comment by jvmvik** [comment]
finally README.md and yard doc looks good to me.

**Comment by jvmvik** [comment]
I might have ran out GitHub Action runtime. This is why the Ruby versions are out of sync and the actions cancelled.

File: README.md [hunk 8]
```diff
@@ -1 +1,600 @@
...
+
+### Search yahoo
+```ruby
+require 'serpapi'
+client = SerpApi::Client.new(api_key: ENV['API_KEY'], engine: 'yahoo')
+results = client.search({
+  "p": "coffee"
+})
+pp results[:organic_results]
+# ENV['API_KEY'] captures the secret user API available from http://serpapi.com
```
Comment (hartator) [inline README.md:192]:
```suggestion
# ENV['API_KEY'] captures the secret user API. You can find yours here: https://serpapi.com/manage-api-key
```

**Review by hartator — CHANGES_REQUESTED** [review]
Thanks @jvmvik, good job. 👍

image

Sorry for the long review, but it's primarily typos and there are, unfortunately, many of them. If I were you, I would review the full thing again, focusing on things solely from an English language perspective. Like the meaning, goal, and consistency of each sentence, as well as check it for typical typos. My English is not great as well,  but using something like Grammarly has been helpful. Do you want a subscription? We have an account with them for the SerpApi team.

File: README.md [hunk 9]
```diff
@@ -1 +1,600 @@
...
+```
+
+[Link to the gem page](https://rubygems.org/gems/serpapi/)
+
+Ruby versions validated by Github action:
+ - 3.1.2
+ - 2.7.5
+ - 2.6.9
+
+see: [Github actions.](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml)
```
[the remark this answers is not in the document]
Comment (jvmvik) [reply README.md:26]: sure. I'll move it.

File: README.md [hunk 10]
```diff
@@ -1 +1,600 @@
...
+
+[![serpapi-ruby](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml/badge.svg)](https://github.com/serpapi/serpapi-ruby/actions/workflows/ci.yml)
+[![Gem Version](https://badge.fury.io/rb/serpapi.svg)](https://badge.fury.io/rb/serpapi)
+
+SerpApi Gem is meant to scrape and parse results from all major search engines including Google, Bing, Baidu, Yandex, Yahoo, Ebay, Apple and more using [SerpApi](https://serpapi.com).
+The service SerpApi.com returns JSON or HTML results. This library does automatically convert JSON in Hash using the json standard library from shipped with Ruby.
+
+## Installation
+
+A modern version of Ruby must be already installed.
```
[the remark this answers is not in the document]
Comment (jvmvik) [reply README.md:13]: fixed.

File: lib/serpapi/serpapi.rb [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    def search(parameter = {})
+      get('/search', :json, parameter)
+    end
+
+    # html search
+    #
+    # @return [String] raw html search results directly from the search engine
+    def html(parameter = {})
+      get('/search', :html, parameter)
+    end
```
[the remark this answers is not in the document]
Comment (jvmvik) [reply lib/serpapi/serpapi.rb:57]: renamed.

File: lib/serpapi/serpapi.rb [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+    # @return [String] default search engine
+    def engine
+      @default['engine'] || @default[:engine]
+    end
+
+    # @return [Hash] parameter default as defined in the constructor
+    def default_parameter
+      @default
+    end
```
[the remark this answers is not in the document]
Comment (jvmvik) [reply lib/serpapi/serpapi.rb:94]: bad copy/paste from the previous implementation. 
I switched.
` attr_accessor :params`
it's evident that the instance of a class is default parameter.
```
client = new SerApi::Client(q: "I write bad code like ChatGPT")
pp client.params[:q].include? 'ChatGPT'
```

File: README.md [hunk 13]
```diff
@@ -1 +1,600 @@
...
+
+```bash
+$ gem install serpapi
```
[the remark this answers is not in the document]
Comment (jvmvik) [reply README.md:31]: done.

**Pull request merged**